onnxruntime/test/contrib_ops/ngram_test.cc

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "gtest/gtest.h"
+#include "test/providers/provider_test_utils.h"
+
+#include <stdint.h>
+
+namespace onnxruntime {
+namespace test {
+namespace ngram_test {
+
+constexpr const char* domain = onnxruntime::kMSDomain;
+const int opset_ver = 1;
+
+void InitTestAttr(OpTester& test, const std::string& mode,
+                  int64_t M, int64_t N, int64_t S, bool all,
+                  const std::vector<int64_t>& ngram_counts,
+                  const std::vector<int64_t>& ngram_indexes,
+                  const std::vector<float>& weights,
+                  const std::vector<int64_t>& pool_int64s,
+                  const std::vector<std::string>& pool_strings) {
+  test.AddAttribute("mode", mode);
+  test.AddAttribute("M", M);
+  test.AddAttribute("N", N);
+  test.AddAttribute("S", S);
+  test.AddAttribute("all", int64_t{all});
+  test.AddAttribute("ngram_counts", ngram_counts);
+  test.AddAttribute("ngram_indexes", ngram_indexes);
+  // optional
+  if (!weights.empty()) {
+    test.AddAttribute("weights", weights);
+  }
+  if (!pool_int64s.empty()) {
+    test.AddAttribute("pool_int64s", pool_int64s);
+  } else {
+    test.AddAttribute("pool_strings", pool_strings);
+  }
+}
+}  // namespace ngram_test
+
+using namespace ngram_test;
+
+TEST(ContribOpTest, Ngram_Int32_TF_AllFalse_Skip0) {
+  OpTester test("Ngram", opset_ver, domain);
+  // 1 - 2, s=0, , all=false, weights empty, int32
+  InitTestAttr(test, "TF", 1, 2, 0, false,
+               {0, 4},
+               {0, 1, 2, 3, 4, 5, 6},  //7 output indexes
+               {},
+               {2, 3, 5, 4,         //1-grams
+                5, 6, 7, 8, 6, 7},  //bi-grams
+               {});
+
+  std::vector<int64_t> dims{2, 6};
+  std::vector<int32_t> input = {1, 1, 3, 3, 3, 7, 8, 6, 7, 5, 6, 8};
+  test.AddInput<int32_t>("T", dims, input);
+
+  std::vector<int64_t> out_dims{7};
+  std::vector<float> output = {0, 0, 0, 0, 1, 1, 1};
```

Comment:
>0, 0, 0 [](start = 34, length = 7)

I expected the counts for the 1-gram '3' be 3 and the 1-gram '5' be 1. Also expected the count for the 2-gram '7,8' to be 0 given the occurence of 7 and 8 is split over the 2 rows of input isn't it?

Can you explain the output as maybe I'm misinterpreting what it should be?

Reply:
I am going to add comments to the unit tests. 
Here is what takes place in general. There are 7 n-grams in the input. 4 unigrams and 3 bigrams that are expressed as 10 items (integers in this case) of input.
We only count and then optionally scale those ngrams that appear in the supplied pool parameter (either int64 or string).
M = 1 and N =2 in this case. However, attribute *all* controls if we take all of the supplied ngrams [M..N] sizes into consideration or not With all=false, we only consider N-grams which case bigrams. This explains why unigrams counts are all zeros.

---
In reply to: [242801303](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r242801303) [](ancestors = 242801303)

Reply:
Based on this explanation, I don't see the point of the 'all' attribute. If we only want output from N, can we not just specify the same value for M and N?

---
In reply to: [243047440](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243047440) [](ancestors = 243047440,242801303)

Reply:
You're right. We can safely remove `all`. `all=true` will be replaced with `M=1` and `N=N`. `all=false` can be encoded as `M=N`.

---
In reply to: [243105284](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r243105284) [](ancestors = 243105284,243047440,242801303)